pkg/liquidity-mining/contracts/admin/ChildChainGaugeTokenAdder.sol

```diff
@@ -0,0 +1,122 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/Authentication.sol";
+
+import "../interfaces/IAuthorizerAdaptor.sol";
+import "../interfaces/IChildChainLiquidityGaugeFactory.sol";
+
+/**
+ * @title ChildChainGaugeTokenAdder
+ * @notice Allows atomically adding a new reward token to a RewardsOnlyGauge while ensuring that it remains in sync
+ * with its ChildChainStreamer.
+ */
+contract ChildChainGaugeTokenAdder is Authentication {
+    // RewardsOnlyGauge expects the claim function selector to be left padded with zeros.
+    // We then shift right 28 bytes so that the function selector (top 4 bytes) sits in the lowest 4 bytes.
+    bytes32 private constant _CLAIM_SIG = keccak256("get_reward()") >> (28 * 8);
+    uint256 private constant _MAX_TOKENS = 8;
+    uint256 private constant _REWARD_DURATION = 1 weeks;
+
+    IVault private immutable _vault;
+    IAuthorizerAdaptor private immutable _authorizerAdaptor;
+    IChildChainLiquidityGaugeFactory private immutable _gaugeFactory;
+
+    constructor(IChildChainLiquidityGaugeFactory gaugeFactory, IAuthorizerAdaptor authorizerAdaptor)
+        Authentication(bytes32(uint256(address(this))))
+    {
+        // ChildChainGaugeTokenAdder is a singleton, so it uses its own address to disambiguate action identifiers.
+
+        _vault = authorizerAdaptor.getVault();
+        _authorizerAdaptor = authorizerAdaptor;
+        _gaugeFactory = gaugeFactory;
+    }
+
+    /**
+     * @notice Returns the Balancer Vault
+     */
+    function getVault() public view returns (IVault) {
+        return _vault;
+    }
+
+    /**
+     * @notice Returns the Balancer Vault's current authorizer.
+     */
+    function getAuthorizer() public view returns (IAuthorizer) {
+        return getVault().getAuthorizer();
+    }
+
+    /**
+     * @notice Returns the address of the Authorizer adaptor contract.
+     */
+    function getAuthorizerAdaptor() external view returns (IAuthorizerAdaptor) {
+        return _authorizerAdaptor;
+    }
+
+    /**
+     * @notice Adds a new token to a RewardsOnlyGauge.
+     */
+    function addTokenToGauge(
+        IRewardsOnlyGauge gauge,
+        IERC20 rewardToken,
+        address distributor
+    ) external authenticate {
+        require(_gaugeFactory.isGaugeFromFactory(address(gauge)), "Invalid gauge");
+        IChildChainStreamer streamer = IChildChainStreamer(_gaugeFactory.getGaugeStreamer(address(gauge)));
+        require(streamer == gauge.reward_contract(), "Not original gauge streamer");
+
+        // We first add the new token to the streamer so that the gauge can claim it when checkpointing.
+        _addTokenToStreamer(streamer, rewardToken, distributor);
```

Comment:
This actually doesn't happen: the gauge checkpoints before and after adding the token, but the second checkpoint does nothing because of the 1 hour maximum frequency.

Reply:
I'm meaning this in the ongoing sense here rather than for the next claim specifically.